### ISSUE TITLE
How to train a Guided-Classifier？

### ISSUE THREAD

**Issue by ONobody** [Derekabc/DMWeeds#4]
The first step of my training was to train diffusion. 
Here are my parameters：
MODEL_FLAGS="--image_size 256 --num_channels 256 --num_res_blocks 2 --learn_sigma True --use_scale_shift_norm true --attention_resolutions 32,16,8 --num_head_channels 64"
DIFFUSION_FLAGS="--resblock_updown True --diffusion_steps 1000 --noise_schedule linear --rescale_learned_sigmas False --rescale_timesteps False "
TRAIN_FLAGS="--lr 1e-4 --batch_size 2 --dropout 0.1"
python scripts/image_train.py --data_dir /tmp/pycharm_project_123/datasets/image $MODEL_FLAGS $DIFFUSION_FLAGS $TRAIN_FLAGS
If I continue to want to train guided_classifier
Do you need to change any codes?
thank you

Add the above MODEL_FALGS to class_cond True
then python classifier_train.py ........  ?

**Comment by ONobody** [comment]
Can you help? Thank you very much.

**Comment by DongChen06** [comment]
Following are the codes I used, also please change the code accordingly to adopt fine-tuning. (https://github.com/DongChen06/guided-diffusion)

# Train Diffusion Model
`MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond True --image_size 256 --learn_sigma True --num_channels 256 --num_heads 4 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"
DIFFUSION_FLAGS="--noise_schedule linear"
TRAIN_FLAGS="--lr 1e-4 --batch_size 2"`
	
`python image_train.py --data_dir ../dataset/cassava_leaf_train  --resume_checkpoint  ../model256/256x256_diffusion.pt $MODEL_FLAGS $DIFFUSION_FLAGS $TRAIN_FLAGS`

![image](https://user-images.githubusercontent.com/25771207/223435558-498bd8c2-6eb7-4bc2-9285-4397ab178439.png)
![image](https://user-images.githubusercontent.com/25771207/223435616-4edf6839-b35b-4582-ad5a-b2308e2118ce.png)


# Sample with diffusion model only.
`MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond True --image_size 256 --learn_sigma True --num_channels 256 --num_heads 4 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"
DIFFUSION_FLAGS="--noise_schedule linear"`

`python image_sample.py --model_path ../model256/ema_0.9999_300000.pt $MODEL_FLAGS $DIFFUSION_FLAGS  --num_samples 2000`

![image](https://user-images.githubusercontent.com/25771207/223435699-40872f4a-6e4b-48fe-9518-1e2ce3707871.png)


# Train Classifier
Here are flags for training the classifier. You can modify these for training classifiers at other resolutions:
`TRAIN_FLAGS="--iterations 300000 --anneal_lr True --batch_size 6 --lr 3e-4 --save_interval 10000 --weight_decay 0.05"
CLASSIFIER_FLAGS="--image_size 256 --classifier_attention_resolutions 32,16,8 --classifier_depth 2 --classifier_width 128 --classifier_pool attention --classifier_resblock_updown True --classifier_use_scale_shift_norm True"`

`python classifier_train.py --data_dir ../dataset/cassava_leaf_train --resume_checkpoint ../model256/256x256_classifier.pt $TRAIN_FLAGS $CLASSIFIER_FLAGS`

![image](https://user-images.githubusercontent.com/25771207/223435756-74ba2bec-334c-4900-8468-caa2dbab2e38.png)


# Sampling
`MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond True --image_size 256 --learn_sigma True --num_channels 256 --num_heads 4 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"`

`CLASSIFIER_FLAGS="--image_size 256 --classifier_attention_resolutions 32,16,8 --classifier_depth 2 --classifier_width 128 --classifier_pool attention --classifier_resblock_updown True --classifier_use_scale_shift_norm True"`

`SAMPLE_FLAGS="--batch_size 32 --num_samples 2000 --timestep_respacing 250"`

`python classifier_sample.py  --model_path ../model256/ema_0.9999_300000.pt  --classifier_path ../model256/model299999.pt $MODEL_FLAGS $CLASSIFIER_FLAGS $SAMPLE_FLAGS`

**Comment by ONobody** [comment]
Thank you very much ! !
When calculating FID
![图片2png](https://user-images.githubusercontent.com/108787020/223608577-8d77c928-28d9-4c0f-aada-6dd24ffcd1dc.png)
, you need to convert your data set into NPZ format, right?
Is it the wrong way for me to convert?

**Comment by DongChen06** [comment]
I think the evaluation output should be npz format. Then you can calculate the FID via some code as follows:
`python evaluator.py ../model256/samples_2000_ADMG.npz ../dataset/cassava_leaf2020.npz`

there are some scripts I used to convert the npz to RGB images: https://github.com/DongChen06/guided-diffusion/tree/master/scripts

**Comment by ONobody** [comment]
I want to convert my data set into npz format and then calculate FID with the generated picture.
But I keep making mistakes.  
Just like above.

**Comment by ONobody** [comment]
Sorry to bother you again.
when i Train Classifier 
![image](https://user-images.githubusercontent.com/108787020/226095074-8455f5fa-7508-4dd6-a092-3b523e5f74b2.png)
![image](https://user-images.githubusercontent.com/108787020/226095088-315afae2-b374-479c-b69b-d5baa8ebf654.png)
But I made such a mistake 
Can you help? I strictly follow your training steps.
